Adding unit tests for SA1308

johnkoerner opened the pull request:
Fixes #318

sharwell commented:
Welcome to the project @johnkoerner!

:bulb: If you put `Fixes #318` in the title, it won't create a link back to the issue automatically. I moved it to the body of the post instead which created the link.

sharwell commented:
:bulb: You'll find that the implementation for the SA1308 diagnostic contains a reference to `AnalyzerConstants.DisabledNoTests`. The person who implements the tests has the first chance to change that value to `true`, which enables the diagnostic for our release builds.

johnkoerner commented:
So, should I change that value after writing the tests or after the tests are merged into master?

sharwell commented:
You aren't required to change it at all (if you don't then one of the others will later). But if you want you can include it either in the same pull request that adds the tests, or in one that follows it.

sharwell commented on StyleCop.Analyzers/StyleCop.Analyzers.Test/NamingRules/SA1308UnitTests.cs:
```diff
@@ -0,0 +1,82 @@
+            {
+                await this.TestFieldSpecifyingModifierAndPrefix(modifier, "m_", "m_");
+                await this.TestFieldSpecifyingModifierAndPrefix(modifier, "s_", "s_");
+                await this.TestFieldSpecifyingModifierAndPrefix(modifier, "m\u005F", "m_");
```
:bulb: Interesting test, but I think you meant `\\u005F`. In it's current form, it's exactly equivalent to the first test.

johnkoerner replied:
Great catch.  I'll fix this.

sharwell commented on StyleCop.Analyzers/StyleCop.Analyzers.Test/NamingRules/SA1308UnitTests.cs:
```diff
@@ -0,0 +1,82 @@
+                await this.TestFieldSpecifyingModifierAndPrefix(modifier, "m_", "m_");
+                await this.TestFieldSpecifyingModifierAndPrefix(modifier, "s_", "s_");
+                await this.TestFieldSpecifyingModifierAndPrefix(modifier, "m\u005F", "m_");
+                await this.TestFieldSpecifyingModifierAndPrefix(modifier, "s\u005F", "s_");
```
:exclamation: This escape sequence is incorrect as well.

sharwell commented:
Hi @johnkoerner,

I completed my review and it's ready to merge aside from the two issues I mentioned.

:bulb: I noticed that the file uses a combination of spaces and hard tabs for indentation. You can fix this when you update the pull request, but it's not required (I'll just correct it when I merge your pull request if it's still there).

sharwell commented:
:bulb: One last thing @johnkoerner - You are not required to update this. Out of courtesy to people who like to update their own work, I made comments here rather than go and change things. If you are busy or for any reason want another person to go ahead and finish these details, please just add a comment to let us know. :+1:

johnkoerner commented:
Thanks for the feedback @sharwell . I fixed the escape sequences and enabled the diagnostic.  I also fixed the tabs and changed my VS setting to the correct setting for tabs vs spaces :).

sharwell commented:
@johnkoerner This is now merged. Please note the following:

1. Thanks for your contribution. The momentum of this project is great and it's all because of people volunteering time to fix things one at a time. :smile: 
2. I merged everything except the merge commit at the end of this pull request, which means GitHub doesn't recognize that this was merged. You *generally* do not need to merge master into your working branch. I highly recommend creating new branches in order to work on new features rather than sending a pull request from the `master` branch of your fork. If you have questions about this process feel free to open a new issue with a question.
3. I made a few insignificant changes during the merge. The primary item you will notice is the simplification of the call to `WithArguments` in `TestFieldSpecifyingModifierAndPrefix`.
